### ISSUE TITLE
模型实现问题

### ISSUE THREAD

**Issue by niningliuhen2013** [nefujiangping/EncAttAgg#5]
首先非常感谢您共享了代码！结合论文看了代码，有几个问题想请教一下：
1.mutual_attender和integration_attender 部分的代码，一个是公式4一个是公式6，代码核心都是实现Transformer的一个Encoder层，为什么mutual_attender没有和integration_attender 一样使用nn.TransformerEncoderLayer呢？
2.models.py中的TwoLayerLinear是实现公式7吗，但没有看到LogSumExp这个公式的实现部分，还是LogSumExp在哪里实现了我没看见呢？

期待您的回复，感谢！

**Comment by nefujiangping** [comment]
1、[nn.TransformerEncoderLayer](https://pytorch.org/docs/master/generated/torch.nn.TransformerEncoderLayer.html#torch.nn.TransformerEncoderLayer) forward 的输入只有一个 src，其内部调用的 `MultiheadAttention(src, src, src)`；也就是Q/K/V是一样的。integration_attender 公式6输入就是一个tensor，就可以直接用`nn.TransformerEncoderLayer`，但是mutual_attender 公式4，输入H和T，只能用 `MultiheadAttention` 自己写。
2、`TwoLayerLinear` 就是公式7的 `W1 -> relu -> W2`；`LogSumExp` 函数在[这里](https://github.com/nefujiangping/EncAttAgg/blob/87386800add9895f2d4d3bb9b6ad0e439ddaf43d/models/models.py#L472)直接调用的torch的 `logsumexp` 函数。

抱歉，回复晚了，希望能解答你的疑惑。

**Comment by niningliuhen2013** [comment]
感谢感谢，后面自己研究代码搞明白了。
不过还有几个问题想要请教一下：
1. 我在gen_data.py中统计了一下DocRed训练集、验证集和测试集的三元组的正样本和负样本的数目，发现几个与论文中的table 1的有出入的数据，训练集的负样本实体对数目1163035，验证集的负样本实体对数目385272，测试集的正样本实体对数目为0，测试集负样本实体对数目为379316，测试集没有labels标签，请问表格里测试集的正样本实体对怎么获得呢
2. data_loader.py中collate_fn_bert_enc函数中head_mentions_indices_mask、tail_mentions_indices_mask、ht_comb_mask、t2p_map_mask添加了dummy mask，这样会引入错误吧，虽然是为了避免出现nan。其中t2p_map_mask的问题可以在Embedding Class的foward()的seq_repr = torch.sum(seq_repr, dim=2) / t2p_map_mask.sum(dim=-1, keepdim=True).float()改为seq_repr = torch.mean(seq_repr, dim=2)；head_mentions_indices_mask、tail_mentions_indices_mask发现则在MutualAttender.ML_MMA的nn.MultiheadAttention出现的问题，ht_comb_mask是在IntegrationAttender.ML_MA的TransformerEncoderLayer的MultiheadAttention出现的问题，可能的原因是https://github.com/pytorch/pytorch/issues/24816，目前还没想到如何解决这个，不知道我的理解对不对，请问您后面有新的更好的解决方法吗

期待您的回复，感谢！
